File: src/main/java/mrk/hackerrank/HackerRank03.java
package mrk.hackerrank;

import java.util.Arrays;
import java.util.function.IntPredicate;

//https://www.hackerrank.com/challenges/plus-minus/problem
public class HackerRank03 {

    private void printLine(double number) {
        System.out.println(String.format("%.4f", number));
    }

    private double calculateNumber(int[] arr, IntPredicate predicate) {
        return Arrays.stream(arr).filter(predicate).count() / (double) arr.length;
    }

    public void plusMinus(int[] arr) {
        printLine(calculateNumber(arr, e -> e > 0));
        printLine(calculateNumber(arr, e -> e < 0));
        printLine(calculateNumber(arr, e -> e == 0));
    }
}

File: kotlin/src/main/kotlin/HackerRank03.kt
import java.text.DecimalFormat

// https://www.hackerrank.com/challenges/plus-minus/problem
fun hackerRank03(arr: Array<Int>): String {
    val format = DecimalFormat("0.000000")

    fun calculateRatio(predicate: (Int) -> Boolean) = arr.count(predicate) / arr.size.toDouble()

    return listOf(
        calculateRatio { it > 0 },
        calculateRatio { it < 0 },
        calculateRatio { it == 0 },
    ).joinToString(separator = "\n") { format.format(it) }
}
